File: dnn-providers/hip-kernel-provider/rocke/platform/cpp/instances/gfx942/tf32_mma_probe.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/* Spec-driven native mirror of instances/gfx942/tf32_mma_probe.py. */
#include "rocke/arch_target.h"
#include "rocke/error_boundary.hpp"
#include "rocke/helper_rocke.helpers.mma_io.h"
#include "rocke/instance_tf32_mma_probe.h"
#include "rocke/ir_internal.h"
#include <stdio.h>
#include <string.h>

rocke_kernel_def_t* rocke_build_tf32_mma_probe(rocke_ir_builder_t* b, int m, const char* mode)
{
    char name[96];
    snprintf(name, sizeof(name), "tf32_probe_%d_%s", m, mode ? mode : "invalid");
    if(rocke_ir_builder_init(b, name) != ROCKE_OK)
        return NULL;
    return ckc::guard_builder(b, [&]() -> rocke_kernel_def_t* {
        if((m != 16 && m != 32) || !mode
           || (strcmp(mode, "raw") && strcmp(mode, "carrier") && strcmp(mode, "rne")
               && strcmp(mode, "prepacked") && strcmp(mode, "fp32")))
            return (rocke_kernel_def_t*)rocke_i_set_err(
                b, ROCKE_ERR_VALUE, "TF32 probe requires m=16/32 and a known preparation");
        int k = 128 / m;
        const rocke_arch_target_t* target = rocke_arch_target_from_gfx("gfx942");
        const rocke_mma_op_t* atom = rocke_mma_catalog_op_for_shape(
            &target->mma, "mma", "tf32", "tf32", "fp32", m, m, k, NULL);
        const rocke_type_t* ty = strcmp(mode, "prepacked") == 0 ? rocke_i32() : rocke_f32();
        rocke_param_opts_t opts = {};
        opts.align = 4;
        opts.align_set = true;
        const rocke_type_t* input_ptr = rocke_ptr_type(b, ty, "global");
        rocke_value_t* a = rocke_b_param(b, "A", input_ptr, &opts);
        rocke_value_t* bb = rocke_b_param(b, "B", input_ptr, &opts);
        const rocke_type_t* float_ptr = rocke_ptr_type(b, rocke_f32(), "global");
        rocke_value_t* cc = rocke_b_param(b, "C", float_ptr, &opts);
        rocke_value_t* d = rocke_b_param(b, "D", float_ptr, &opts);
        const rocke_type_t* bits_ptr = rocke_ptr_type(b, rocke_i32(), "global");
        rocke_value_t* pa = rocke_b_param(b, "PA", bits_ptr, &opts);
        rocke_value_t* pb = rocke_b_param(b, "PB", bits_ptr, &opts);
        rocke_value_t* lane = rocke_b_thread_id_x(b);
        rocke_value_t* batch = rocke_b_block_id_x(b);
        rocke_value_t* cm = rocke_b_const_i32(b, m);
        rocke_value_t* axis = rocke_b_mod(b, lane, cm);
        rocke_value_t* group = rocke_b_div(b, lane, cm);
        rocke_value_t* ck = rocke_b_const_i32(b, k);
        rocke_value_t* row = rocke_b_mul(b, axis, ck);
        rocke_value_t* size = rocke_b_const_i32(b, m * k);
        rocke_value_t* batch_input = rocke_b_mul(b, batch, size);
        rocke_value_t* row_base = rocke_b_add(b, batch_input, row);
        size = rocke_b_const_i32(b, m * m);
        rocke_value_t* batch_output = rocke_b_mul(b, batch, size);
        rocke_value_t* indices[16];
        rocke_value_t* c_values[16];
        for(int slot = 0; slot < atom->c_frag_len; ++slot)
        {
            rocke_value_t *r, *col;
            rocke_layout_map_coord(atom->c_layout, b, lane, slot, &r, &col);
            rocke_value_t* offset = rocke_b_mul(b, r, cm);
            offset = rocke_b_add(b, offset, col);
            indices[slot] = rocke_b_add(b, batch_output, offset);
            c_values[slot] = rocke_b_global_load(b, cc, indices[slot], rocke_f32(), 4);
        }
        rocke_value_t* acc = rocke_b_vec_pack(b, c_values, atom->c_frag_len, rocke_f32());
        if(strcmp(mode, "fp32") == 0)
        {
            const rocke_mma_op_t* full = rocke_mma_catalog_op_for_shape(
                &target->mma, "mma", "fp32", "fp32", "fp32", m, m, k / 2, NULL);
            for(int step = 0; step < 2; ++step)
            {
                rocke_value_t* delta = rocke_b_const_i32(b, step * (64 / m));
                delta = rocke_b_add(b, group, delta);
                rocke_value_t* index = rocke_b_add(b, row_base, delta);
                rocke_value_t* av = rocke_b_global_load(b, a, index, rocke_f32(), 4);
                rocke_value_t* bv = rocke_b_global_load(b, bb, index, rocke_f32(), 4);
                rocke_value_t* bits = rocke_b_bitcast(b, av, rocke_i32());
                rocke_b_global_store(b, pa, index, bits, 4);
                bits = rocke_b_bitcast(b, bv, rocke_i32());
                rocke_b_global_store(b, pb, index, bits, 4);
                acc = rocke_b_mma(b, full->op_id, av, bv, acc, NULL, 0);
            }
        }
        else
        {
            rocke_matrix_fragment_layout_t layout = {{{32, 32}, 2, 32, 2}, 2, 64 / m, m};
            rocke_value_t* inputs[] = {a, bb};
            rocke_value_t* prepared[] = {pa, pb};
            rocke_value_t* fragments[2];
            for(int role = 0; role < 2; ++role)
            {
                bool packed = strcmp(mode, "prepacked") == 0;
                const rocke_type_t* carrier
                    = packed || strcmp(mode, "carrier") == 0 ? rocke_i32() : rocke_f32();
                rocke_value_t* values = rocke_h_load_matrix_fragment(b,
                                                                     inputs[role],
                                                                     row_base,
                                                                     group,
                                                                     0,
                                                                     packed ? "tf32" : "fp32",
                                                                     &layout,
                                                                     carrier,
                                                                     4);
                if(strcmp(mode, "rne") == 0)
                {
                    rocke_value_t* elems[2];
                    for(int i = 0; i < 2; ++i)
                    {
                        rocke_value_t* e = rocke_b_vec_extract(b, values, i);
                        elems[i] = rocke_b_cvt_f32_to_tf32(b, e);
                    }
                    values = rocke_b_vec_pack(b, elems, 2, rocke_tf32());
                }
                else
                    values = rocke_b_bitcast(b, values, rocke_vector_type(b, rocke_tf32(), 2));
                for(int slot = 0; slot < 2; ++slot)
                {
                    rocke_value_t* two = rocke_b_const_i32(b, 2);
                    rocke_value_t* offset = rocke_b_mul(b, group, two);
                    rocke_value_t* index = rocke_b_const_i32(b, slot);
                    offset = rocke_b_add(b, offset, index);
                    index = rocke_b_add(b, row_base, offset);
                    rocke_value_t* e = rocke_b_vec_extract(b, values, slot);
                    rocke_value_t* bits = rocke_b_bitcast(b, e, rocke_i32());
                    rocke_b_global_store(b, prepared[role], index, bits, 4);
                }
                fragments[role] = values;
            }
            acc = rocke_b_mma(b, atom->op_id, fragments[0], fragments[1], acc, NULL, 0);
        }
        for(int slot = 0; slot < atom->c_frag_len; ++slot)
        {
            rocke_value_t* value = rocke_b_vec_extract(b, acc, slot);
            rocke_b_global_store(b, d, indices[slot], value, 4);
        }
        rocke_b_ret(b);
        return rocke_ir_builder_kernel(b);
    });
}

File: dnn-providers/hip-kernel-provider/rocke/platform/python/rocke/instances/gfx942/tf32_mma_probe.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
"""One wave per dense atom, with explicit FP32 input-preparation variants."""

from dataclasses import dataclass

from rocke.core.arch import ArchTarget
from rocke.core.ir import F32, I32, TF32, IRBuilder, PtrType, VectorType
from rocke.core.storage import BitPacking, FragmentPacking, MatrixFragmentLayout
from rocke.helpers.mma_io import load_matrix_fragment

PREPARATIONS = ("raw", "carrier", "rne", "prepacked", "fp32")


@dataclass(frozen=True)
class Tf32MmaProbeSpec:
    m: int = 16
    preparation: str = "raw"

    def __post_init__(self):
        if self.m not in (16, 32) or self.preparation not in PREPARATIONS:
            raise ValueError("TF32 probe requires m=16/32 and a known preparation")

    @property
    def name(self):
        return f"tf32_probe_{self.m}_{self.preparation}"


def build_tf32_mma_probe(spec: Tf32MmaProbeSpec):
    m, mode = spec.m, spec.preparation
    k = 128 // m
    target = ArchTarget.from_gfx("gfx942")
    atom = target.mma.op_for_shape(
        family="mma",
        a_dtype="tf32",
        b_dtype="tf32",
        c_dtype="fp32",
        m=m,
        n=m,
        k=k,
    )
    b = IRBuilder(spec.name)
    ty = I32 if mode == "prepacked" else F32
    a = b.param("A", PtrType(ty, "global"), align=4)
    bb = b.param("B", PtrType(ty, "global"), align=4)
    cc = b.param("C", PtrType(F32, "global"), align=4)
    d = b.param("D", PtrType(F32, "global"), align=4)
    pa = b.param("PA", PtrType(I32, "global"), align=4)
    pb = b.param("PB", PtrType(I32, "global"), align=4)
    lane = b.thread_id_x()
    batch = b.block_id_x()
    cm = b.const_i32(m)
    axis = b.mod(lane, cm)
    group = b.div(lane, cm)
    ck = b.const_i32(k)
    row = b.mul(axis, ck)
    batch_input = b.mul(batch, b.const_i32(m * k))
    row_base = b.add(batch_input, row)
    batch_output = b.mul(batch, b.const_i32(m * m))
    c_values, out_indices = [], []
    for slot in range(atom.c_frag_len):
        r, col = atom.c_layout().coord(b, lane, slot)
        index = b.add(batch_output, b.add(b.mul(r, cm), col))
        out_indices.append(index)
        c_values.append(b.global_load(cc, index, F32, align=4))
    acc = b.vec_pack(c_values, F32)
    if mode == "fp32":
        full = target.mma.op_for_shape(
            family="mma",
            a_dtype="fp32",
            b_dtype="fp32",
            c_dtype="fp32",
            m=m,
            n=m,
            k=k // 2,
        )
        for step in range(2):
            index = b.add(row_base, b.add(group, b.const_i32(step * (64 // m))))
            av = b.global_load(a, index, F32, align=4)
            bv = b.global_load(bb, index, F32, align=4)
            b.global_store(pa, index, b.bitcast(av, I32), align=4)
            b.global_store(pb, index, b.bitcast(bv, I32), align=4)
            acc = b.mma(full, av, bv, acc)
    else:
        layout = MatrixFragmentLayout(
            FragmentPacking(BitPacking(32), 2, 32, 2), 2, 64 // m, m
        )
        fragments = []
        for ptr, prepared in ((a, pa), (bb, pb)):
            carrier = I32 if mode in ("carrier", "prepacked") else F32
            values = load_matrix_fragment(
                b,
                ptr,
                row_base,
                group,
                0,
                dtype="tf32" if mode == "prepacked" else "fp32",
                layout=layout,
                carrier_type=carrier,
                alignment_bytes=4,
            )
            if mode == "rne":
                values = b.vec_pack(
                    [b.cvt_f32_to_tf32(b.vec_extract(values, i)) for i in range(2)],
                    TF32,
                )
            else:
                values = b.bitcast(values, VectorType(TF32, 2))
            for slot in range(2):
                index = b.add(
                    row_base, b.add(b.mul(group, b.const_i32(2)), b.const_i32(slot))
                )
                bits = b.bitcast(b.vec_extract(values, slot), I32)
                b.global_store(prepared, index, bits, align=4)
            fragments.append(values)
        acc = b.mma(atom, *fragments, acc)
    for slot, index in enumerate(out_indices):
        b.global_store(d, index, b.vec_extract(acc, slot), align=4)
    b.ret()
    return b.kernel
